pkg/net/network.go

```diff
@@ -234,77 +138,33 @@ func (n *network) Listen(ctx context.Context, address string) (chan *Connection,
 	}
 
 	logger.Info("Started listening", zap.Strings("addresses", addresses))
-	n.addAddress(addresses...)
-	n.addAddress(n.relayAddresses...)
+	n.local.AddAddress(addresses...)
+	n.local.AddAddress(n.relayAddresses...)
 
 	cconn := make(chan *Connection, 10)
 	go func() {
 
 		for {
 			tcpConn, err := tcpListener.Accept()
 			if err != nil {
-				log.DefaultLogger.Warn("could not accept connection", zap.Error(err))
+				log.DefaultLogger.Warn(
+					"could not accept connection", zap.Error(err))
 				// TODO close conn?
 				return
 			}
 
 			conn := &Connection{
 				Conn:          tcpConn,
 				RemotePeerKey: nil,
+				IsIncoming:    true,
 			}
 
-			synObj, err := Read(conn)
+			for _, mh := range n.middleware {
+				conn, err = mh(ctx, conn)
 			if err != nil {
-				log.DefaultLogger.Warn("waiting for syn failed", zap.Error(err))
-				// TODO close conn?
-				continue
+					log.DefaultLogger.Error(
+						"middleware failure", zap.Error((err)))
```

Comment:
If a middleware errors I think we should stop processing, error out, and close the connection.
Also if a connection returns nil as the connection, we should also stop processing and close the connection.

Reply:
@geoah Can you check this now?